### PR TITLE
Distance metric refinement

### PR DESCRIPTION
**Pull request by chrisholder**
#### What does this implement/fix? Explain your changes.

TLDR:
Improves the speed of the metrics and fixes a bug that occurred when using the numba fastmath flag and bounding matrix parameters. In addition the lower_bounding parameter has been removed and instead the lower_bounding technique to use is inferred based on the other parameters passed.

Full version:
Previously the distances weren't being cache'd as efficiently as possible and so the cache flag has been added to the factory methods to allow them to be cache'd for repeat calls.

In addition there was a bug caused by using the fastmath numba flag where values were being inserted into the wrong indexes of the bounding matrix (still not sure why this is caused). As such for now the fastmath flag has been removed.

Finally the lower_bounding parameter felt redundent as it could simply be inferred by which parameters are being passed and therefore the parameter was removed.

##### For all contributions
- [x] I've added myself to the [list of contributors](https://github.com/alan-turing-institute/sktime/blob/main/.all-contributorsrc).
- [x] Optionally, I've updated sktime's [CODEOWNERS](https://github.com/alan-turing-institute/sktime/blob/main/CODEOWNERS) to receive notifications about future changes to these files.
- [x] I've added unit tests and made sure they pass locally.

##### For new estimators
- [x] I've added the estimator to the online documentation.
- [x] I've updated the existing example notebooks or provided a new one to showcase how my estimator works.

### CODE REVIEW

**Review by fkiraly — CHANGES_REQUESTED** [review]
I think this is ok if we can get it in for the next release.
If the release happens before the merge, I would have major reservations because this would result in interface changes, so can we get this done quickly?

Besides that, I only have one minor comment: please remove any commented out code.

**Comment by TonyBagnall** [comment]
> I think this is ok if we can get it in for the next release. If the release happens before the merge, I would have major reservations because this would result in interface changes, so can we get this done quickly?
> 
> Besides that, I only have one minor comment: please remove any commented out code.

hi franz, what interface changes are there? All of thes numba distances are new, and all the existing code is still there

**Review by TonyBagnall — DISMISSED** [review]
I have tested this for correctness and speed on dtw. Any further problems can be resolved by latter PRs, but I am confident this is now all correct.

**Review by TonyBagnall — APPROVED** [review]
